### PR TITLE
Refactor endpoint setup in acceptance tests

### PR DESCRIPTION
**Pull request by timbussmann**
Refactors the endpoint templates for setting up endpoints in acceptance tests. The two major changes are:

* `ExternallyManagedContainerServer` seems no longer be doing anything useful in v8, so it's being removed and every remaining logic is moved into `DefaultServer`. `DefaultServer` now inherits from `ServerWithNoDefaultPersistenceDefinitions` which handles the transport setup while `DefaultServer` also provides the persistence setup part.
* `builder.TypesToIncludeInScan(endpointConfiguration.GetTypesScopedByTestClass());` is moved to the end of the configuration. This came up from the recent changes that disabled file-based assembly scanning for acceptance tests. This requires all necessary assemblies to scan to be loaded into the AppDomain when the scanning takes part. Some downstreams, primarily persisters, had issues loading all persistence related features because they were not picked up by scanning at the beginning (as the definition of the transport that causes the assembly to be loaded happens after the scanning).

### CODE REVIEW

**Review by HEskandari — APPROVED** [review]

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/ServerWithNoDefaultPersistenceDefinitions.cs [hunk 1]
```diff
@@ -1,48 +1,34 @@
...
 
-            await builder.DefineTransport(TestSuiteConstraints.Current.CreateTransportConfiguration(), runDescriptor, endpointConfiguration).ConfigureAwait(false);
+            await builder.DefineTransport(TransportConfiguration, runDescriptor, endpointConfiguration).ConfigureAwait(false);
 
             builder.RegisterComponentsAndInheritanceHierarchy(runDescriptor);
 
-            configurationBuilderCustomization(builder);
+            await configurationBuilderCustomization(builder).ConfigureAwait(false);
+
+            // scan types at the end so that all types used by the configuration have been loaded into the AppDomain
```
Comment (mikeminutillo) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/ServerWithNoDefaultPersistenceDefinitions.cs:28]: This comment should probably also be added to the `AcceptanceTestingTransportServer`

**Review by mikeminutillo — APPROVED** [review]

**Review by andreasohlund — APPROVED** [review]

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs [hunk 2]
```diff
@@ -4,13 +4,17 @@
...
+
+        public override Task<EndpointConfiguration> GetConfiguration(RunDescriptor runDescriptor, EndpointCustomizationConfiguration endpointConfiguration, Func<EndpointConfiguration, Task> configurationBuilderCustomization)
         {
-            return base.GetConfiguration(runDescriptor, endpointCustomizationConfiguration, endpointConfiguration =>
+            return base.GetConfiguration(runDescriptor, endpointConfiguration, async configuration =>
             {
-                configurationBuilderCustomization(endpointConfiguration);
+                await configuration.DefinePersistence(PersistenceConfiguration, runDescriptor, endpointConfiguration).ConfigureAwait(false);
+
+                await configurationBuilderCustomization(configuration);
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs:17]:
```suggestion
                await configurationBuilderCustomization(configuration).ConfigureAwait(false);
```

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultPublisher.cs [hunk 3]
```diff
@@ -6,9 +6,6 @@ namespace NServiceBus.AcceptanceTests.EndpointTemplates
 
     public class DefaultPublisher : IEndpointSetupTemplate
     {
-        public Task<EndpointConfiguration> GetConfiguration(RunDescriptor runDescriptor, EndpointCustomizationConfiguration endpointConfiguration, Action<EndpointConfiguration> configurationBuilderCustomization)
-        {
-            return new DefaultServer().GetConfiguration(runDescriptor, endpointConfiguration, configurationBuilderCustomization);
-        }
+        public Task<EndpointConfiguration> GetConfiguration(RunDescriptor runDescriptor, EndpointCustomizationConfiguration endpointConfiguration, Func<EndpointConfiguration, Task> configurationBuilderCustomization) => new DefaultServer().GetConfiguration(runDescriptor, endpointConfiguration, configurationBuilderCustomization);
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultPublisher.cs:9]: Personal preference but I would actually wrap the long line. Feel free to ignore

File: src/NServiceBus.AcceptanceTests/Core/SubscriptionMigration/AcceptanceTestingTransportServer.cs [hunk 4]
```diff
@@ -39,8 +35,9 @@ public async Task<EndpointConfiguration> GetConfiguration(RunDescriptor runDescr
             await persistenceConfiguration.Configure(endpointConfiguration.EndpointName, configuration, runDescriptor.Settings, endpointConfiguration.PublisherMetadata);
             runDescriptor.OnTestCompleted(_ => persistenceConfiguration.Cleanup());
 
-            configurationBuilderCustomization(configuration);
+            await configurationBuilderCustomization(configuration);
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Core/SubscriptionMigration/AcceptanceTestingTransportServer.cs:38]:
```suggestion
            await configurationBuilderCustomization(configuration).ConfigureAwait(false);
```

Comment (danielmarbach) [reply src/NServiceBus.AcceptanceTests/Core/SubscriptionMigration/AcceptanceTestingTransportServer.cs:38]: Or consistently remove it if we generally don add it in the ATTs. I don know what the latest state is there

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs:17]: Or consistently remove it if we generally don add it in the ATTs. I don know what the latest state is there

**Review by danielmarbach — APPROVED** [review]
Minor suggestions

File: src/NServiceBus.AcceptanceTests/Core/SubscriptionMigration/AcceptanceTestingTransportServer.cs (hunk 4 above)
Comment (timbussmann) [reply src/NServiceBus.AcceptanceTests/Core/SubscriptionMigration/AcceptanceTestingTransportServer.cs:38]: it seems ConfigureAwait is generally not used, so I removed it

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/ServerWithNoDefaultPersistenceDefinitions.cs (hunk 1 above)
Comment (timbussmann) [reply src/NServiceBus.AcceptanceTests/EndpointTemplates/ServerWithNoDefaultPersistenceDefinitions.cs:28]: done

**Pull request merged**